src-self-hosted/codegen.zig

```diff
@@ -348,172 +350,182 @@ const Function = struct {
+                            //
+                            // From section 2.2.1.2 of the manual, REX is encoded as b0100WRXB. In this case, that's
+                            // b01000101, or 0x45.
+                            return self.code.appendSlice(&[_]u8{
+                                0x45,
+                                0x31,
+                                0xC0 | (@intCast(u8, @truncate(u3, reg.id())) << 3) | @truncate(u3, reg.id()),
+                            });
+                        } else {
+                            return self.code.appendSlice(&[_]u8{
+                                0x31,
+                                0xC0 | (@intCast(u8, reg.id()) << 3) | @intCast(u3, reg.id()),
+                            });
                         }
-                        // Next best case: set eax with 4 bytes
-                        // b8 04 03 02 01           mov    eax,0x01020304
-                        if (x <= std.math.maxInt(u32)) {
+                    }
+                    if (x <= std.math.maxInt(u32)) {
+                        // Next best case: if we set the lower four bytes, the upper four will be zeroed.
+                        //
+                        // The encoding for `mov IMM32 -> REG` is (0xB8 + R) IMM.
+                        if (reg.isExtended()) {
+                            // Just as with XORing, we need a REX prefix. This time though, we only
+                            // need the B bit set, as we're extending the opcode's register field,
+                            // and there is no Mod R/M byte.
+                            //
+                            // Thus, we need b01000001, or 0x41.
+                            try self.code.resize(self.code.items.len + 6);
+                            self.code.items[self.code.items.len - 6] = 0x41;
+                        } else {
                             try self.code.resize(self.code.items.len + 5);
-                            self.code.items[self.code.items.len - 5] = 0xb8;
-                            const imm_ptr = self.code.items[self.code.items.len - 4 ..][0..4];
-                            mem.writeIntLittle(u32, imm_ptr, @intCast(u32, x));
-                            return;
                         }
-                        // Worst case: set rax with 8 bytes
-                        // 48 b8 08 07 06 05 04 03 02 01    movabs rax,0x0102030405060708
-                        try self.code.resize(self.code.items.len + 10);
-                        self.code.items[self.code.items.len - 10] = 0x48;
-                        self.code.items[self.code.items.len - 9] = 0xb8;
-                        const imm_ptr = self.code.items[self.code.items.len - 8 ..][0..8];
-                        mem.writeIntLittle(u64, imm_ptr, x);
+                        self.code.items[self.code.items.len - 5] = 0xB8 | @intCast(u8, @truncate(u3, reg.id()));
```

Comment:
s/intCast/as/

Reply:
and then... why do you need the `@truncate`?
Why not: `0xB8 | @as(u8, reg.id())`

Reply:
Truncate is absolutely necessary. Only three bits of id can be added (notice that the lower byte is already 8), the fourth bit goes into the REX. The truncate removes the potential fourth bit, and it is then cast back to u8 so it `or`s properly

Reply:
... though it could just be `@as(u8, reg.id() & 0b111)`, now that I think about it.

Reply:
oh I thought we were already inside a `isExtended` check here.

Reply:
Not in this instance, nope :)